### PR TITLE
[CI] Add FPGA runner

### PR DESCRIPTION
**Pull request by engdoreis**
This PR adds the FPGA test to the repo using the FPGA runner setup by @nbdd0121.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -75,3 +75,58 @@ jobs:
...
+      uses: cachix/install-nix-action@v27
+      with:
+        extra_nix_config: |
+          substituters = https://nix-cache.lowrisc.org/public/ https://cache.nixos.org/
+          trusted-public-keys = nix-cache.lowrisc.org-public-1:O6JLD0yXzaJDPiQW1meVu32JIDViuaPtGDfjlOopU7o= cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY=
+
+    - name: Check bitstream cache
+      run: |
+        BITSTREAM_HASH=$(nix eval .#sonata-simulator --apply 'd: d.outPath' --raw | cut -d '/' -f4 | cut -d '-' -f1)
+        if gsutil stat "${GS_PATH}/$BITSTREAM_HASH.bit" >/dev/null 2>&1; then
```
Comment (nbdd0121) [inline .github/workflows/ci.yml:104]:
```suggestion
        if gcloud storage stat "${GS_PATH}/$BITSTREAM_HASH.bit" >/dev/null 2>&1; then
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -75,3 +75,58 @@ jobs:
...
+          gcould storage cp -v "${GS_PATH}/$BITSTREAM_HASH.bit" "${BITSTREAM_PATH}" 2>/dev/null
```
Comment (nbdd0121) [inline .github/workflows/ci.yml:105]:
```suggestion
          gcloud storage cp -v "${GS_PATH}/$BITSTREAM_HASH.bit" "${BITSTREAM_PATH}" 2>/dev/null
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -75,3 +75,58 @@ jobs:
...
+      if: ${{hashFiles('${BITSTREAM_PATH}') == ''}} 
+      run: |
+        module load xilinx/vivado
+        nix run .#bitstream-build
+
+    - name: Upload bitstream to the cache
+      if: always() && github.event_name != 'pull_request' && steps.build_bitstream.conclusion != 'skipped'
+      run: |
+        BITSTREAM_HASH=$(nix eval .#sonata-simulator --apply 'd: d.outPath' --raw | cut -d '/' -f4 | cut -d '-' -f1)
+        gcould storage cp "${BITSTREAM_PATH}" "${GS_PATH}/$BITSTREAM_HASH.bit"
```
Comment (nbdd0121) [inline .github/workflows/ci.yml:122]:
```suggestion
        gcloud storage cp "${BITSTREAM_PATH}" "${GS_PATH}/$BITSTREAM_HASH.bit"
```

**Review by nbdd0121 — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -75,3 +75,58 @@ jobs:
...
+  fpga:
+    runs-on: [ubuntu-22.04-fpga, sonata]
+    env:
+      BITSTREAM_PATH: build/lowrisc_sonata_system_0/synth-vivado/lowrisc_sonata_system_0.bit
+      GS_PATH: gs://lowrisc-ci-cache/lowRISC/sonata-system/bitstream/
+
+    steps:
+    - uses: actions/checkout@v4
+    - uses: google-github-actions/auth@v2
+      if: always() && github.event_name != 'pull_request'
```
Comment (nbdd0121) [inline .github/workflows/ci.yml:88]: Why `always()`?

File: nix/bitstream.nix [hunk 5]
```diff
@@ -0,0 +1,26 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+{
+  pkgs,
+  lrPkgs,
+  pythonEnv,
+}: {
+  build = pkgs.writeShellApplication {
```
Comment (marnovandermaas) [inline nix/bitstream.nix:9]: You need https://github.com/lowRISC/sonata-system/blob/6ceaa9c298e321eba679d71388da0d6d3e8cd187/flake.nix#L146-L147

Comment (engdoreis) [reply nix/bitstream.nix:9]: Why this is needed? `writeShellApplication` does not have these arguments like derivation.

Comment (nbdd0121) [reply nix/bitstream.nix:9]: These are passed as environment variables into fusesoc

Comment (engdoreis) [reply nix/bitstream.nix:11]: fixed.

File: flake.nix [hunk 6]
```diff
@@ -5,6 +5,7 @@
   description = "Sonata System";
   inputs = {
     lowrisc-nix.url = "github:lowRISC/lowrisc-nix";
+    lowrisc-it.url = "git+ssh://git@github.com/lowrisc/lowrisc-it";
```
Comment (marnovandermaas) [inline flake.nix]: This won't be visible outside of lowRISC.

**Review by marnovandermaas — COMMENTED** [review]
Running local test now. Can you also add some documentation on how to build bitstreams? Maybe just put the command `nix run .#bitstream-build` where we describe bitstream building as an alternative.

File: flake.nix (hunk 6 above)
Comment (marnovandermaas) [reply flake.nix]: already removed.

**Review by marnovandermaas — APPROVED** [review]
This works locally for me, thanks. It would be nice to add some documentation before merging.

**Comment by engdoreis** [comment]
>This works locally for me, thanks. It would be nice to add some documentation before merging.

Added a new commit with the documentation.

File: flake.nix [hunk 7]
```diff
@@ -165,39 +165,25 @@
...
-              "as the first argument."
-            exit 2
-          fi
-          ${./util/test_runner.py} -t 30 fpga "$1" \
-            --elf-file ${sonata-system-software}/bin/test_runner \
-            --tcl-file ${./util/sonata-openocd-cfg.tcl}
-          ${./util/test_runner.py} -t 600 fpga "$1" \
-            --uf2-file ${cheriot-rtos-test-suite}/share/test-suite.uf2
-        '';
+      bitstream = import nix/bitstream.nix {
```
Comment (HU90m) [inline flake.nix:169]: I think this is one commit to early.

File: nix/bitstream.nix [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+# SPDX-License-Identifier: Apache-2.0
+{
+  pkgs,
+  lrPkgs,
+  pythonEnv,
+  FLAKE_GIT_COMMIT,
+  FLAKE_GIT_DIRTY,
+}: {
+  build = pkgs.writeShellApplication {
+    name = "build";
```
Comment (HU90m) [inline nix/bitstream.nix:12]: Could these have slightly more useful names e.g.

```suggestion
    name = "sonata-bitstream-build";
```

**Review by HU90m — CHANGES_REQUESTED** [review]
Nice! Thanks for organising the nix too

File: flake.nix (hunk 7 above)
Comment (engdoreis) [reply flake.nix:169]: Good catch, git absorb sometimes mess think up.

**Review by HU90m — APPROVED** [review]

File: nix/bitstream.nix (hunk 8 above)
Comment (engdoreis) [reply nix/bitstream.nix]: Done

File: doc/dev/fpga-development.md [hunk 9]
```diff
@@ -26,4 +26,10 @@ You can also manually set the initial value of the SRAM, for example:
 ```sh
 fusesoc --cores-root=. run --target=synth --setup --build lowrisc:sonata:system --SRAMInitFile=$PWD/sw/cheri/build/tests/uart_check.vmem
 ```
+## Build bitstream using nix
+Optionally, the bitstream can be built using a nix command:
+```sh
+nix run .#bitstream-build
+``
+Note: Vivado must in one's path for this command to work.
```
Comment (HU90m) [inline doc/dev/fpga-development.md:34]:
````
```
Note: Vivado must be in one's path for this command to work.
````

**Pull request merged**